### PR TITLE
add JTensor, fix #2

### LINKED ISSUE #2: Add helper for TF predict results
Right now TF predict library returns a List of Tensors in the order of `fetch` operations, this is error prone, we should provide a helper that given a list of fetch ops, should return a map (?) of `op name` -> `Tensor`.

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import java.nio.IntBuffer;
+import java.nio.LongBuffer;
+
+@AutoValue
+public abstract class JTensor {
+  static JTensor create(Tensor<?> tensor) {
+    AutoValue_JTensor jt;
+
+    switch (tensor.dataType()) {
+      case INT32:
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: No `BOOL`, `UINT8`, `STRING`?

Comment (ravwojdyla) [reply zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: `STRING` could be `ByteBuffer`, the other `BOOL`, `UINT8` are probably ok to skip.

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import javax.annotation.Nullable;
+import java.nio.DoubleBuffer;
+import java.nio.FloatBuffer;
+import java.nio.IntBuffer;
+import java.nio.LongBuffer;
+
+@AutoValue
+public abstract class JTensor {
+  static JTensor create(Tensor<?> tensor) {
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: do you have to make this method `public`?

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/Utils.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import com.google.common.collect.Maps;
+import org.tensorflow.Session;
+import org.tensorflow.Tensor;
+
+import java.util.List;
+import java.util.Map;
+
+public class Utils {
+
+  public static Map<String, JTensor> runAndExtract(Session.Runner runner, String ...ops) {
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/Utils.java]: `ops` -> `fetchOps`

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/Utils.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    for (String op : ops) {
+      runner.fetch(op);
+    }
+    List<Tensor<?>> tensors = runner.run();
+    Map<String, JTensor> result = Maps.newLinkedHashMap();
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/Utils.java]: `newLinkedHashMapWithExpectedSize`?

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        tensor.close();
+        throw new IllegalStateException("Unsupported data type " + tensor.dataType());
+    }
+
+    tensor.close();
+    return jt;
+  }
+
+  abstract DataType dataType();
+  abstract long[] shape();
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: Could we also add `numDimensions`?

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java (hunk 2 above)
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: + `final` to `tensor`

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/Utils.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.google.common.collect.Maps;
+import org.tensorflow.Session;
+import org.tensorflow.Tensor;
+
+import java.util.List;
+import java.util.Map;
+
+public class Utils {
+
```
Comment (ravwojdyla) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/Utils.java]: \+ private constructor

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java (hunk 2 above)
Comment (nevillelyh) [reply zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: It can be, outside of `Utils#runAndExtract`. I'll make it public?

Comment (ravwojdyla) [reply zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: well - it's a question - i'm not sure how useful this is gonna be to make it public, technically ppl should just go for `Utils#runAndExtract`

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java (hunk 1 above)
Comment (nevillelyh) [reply zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: Is `UINT8` equivalent to `ByteBuffer`?

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java (hunk 2 above)
Comment (regadas) [reply zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: i think we should keep `protected` for now.

File: zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java [hunk 7]
```diff
@@ -74,9 +75,32 @@ static JTensor create(Tensor<?> tensor) {
...
   abstract DataType dataType();
+  abstract int numDimensions();
   abstract long[] shape();
-  @Nullable abstract int[] intValue();
-  @Nullable abstract long[] longValue();
-  @Nullable abstract float[] floatValue();
-  @Nullable abstract double[] doubleValue();
+  abstract Object data();
+
+  public String stringValue() {
```
Comment (regadas) [inline zoltar-tensorflow/src/main/java/com/spotify/zoltar/tf/JTensor.java]: I like this approach 👍

**Comment by regadas** [comment]
lgtm just missing javadoc 😄

**Comment by regadas** [comment]
@nevillelyh can you rebase?

**Pull request merged**